### ISSUE TITLE
Don't generate an exception when an arbitrary symbol is used inside a formula

### ISSUE THREAD

**Issue by YakovL** [Khan/KaTeX#380]
Hi Kevin and thanks for your work first;

current behaviour of KaTeX is such that most of the symols aside latin letters, numbers and special symbols cause `ParseError`. Couple of examples:

1. a formula with dimensions of quantities, in say, Russian, like `{1 Н\over 2 м^2} = 0,5 Па` (in English that would be: `{1 N\over m^2} = 0.5 Pa`)
2. a formula with an "already parsed" symbol like `A = ρgh` (this would be nice for TiddlyWiki as the editable content is this case is more readable)

It would be nice if those symbols are displayed "as is" (except for the font size). In fact, #379 is about this, too.

**Comment by kevinbarabash** [comment]
KaTeX is a group effort.  I think it's important to recognize everyone's contributions.

Just passing certain symbols though without having proper metrics might work in some case, but there's lots of cases where it wouldn't work.  We need to measure the height when doing vertical layouts but if the math contains only characters where we don't have the height we won't be able to create the proper layout.

**Comment by YakovL** [comment]
Yeah, sure, thanks to all other contributors.

> if the math contains only characters where we don't have the height

this sounds unlikely, at least in those cases that I've mentioned. Moreover, I think in real cases either latin letters, digits or simple symbols like `+` (or at least `=`, `>`, `~`, `\rightarrow` or another "expression-making" symbol) will always be present in the base line. Even in an expession like
```
γ = α^β
```
there's `=`. So if I get you right, the height measure doesn't seem that problematic.

**Comment by kevinbarabash** [comment]
If you put `\sqrt{α^β}` we won't know how tall to make the square root.

**Comment by YakovL** [comment]
Ah, you're right, in inline expressions that can be the case..

**Comment by YakovL** [comment]
So, basicaly when a supported symbol is used, its height is calced from the known symbol size? What's the basic length? How the size is made fit to the non-math environment? (if it is) Is it a problem to set the size of α or β the same as an ordinary letter ("a")?

Sorry, as I don't know technical details, I'm not sure what are the right questions to ask, neither I understand in what units different sizes are measured.

**Comment by YakovL** [comment]
As an additional option: there may be a setting that sais "make the size of these symbols equal to that of 'a'". For instance, for Greek alphabet that may be either "αβγδεζηθικλμνξοπρσςτυφχψω" or "αεικνοπστυω", depending on what we have to understand as "the same size" ("δθλ" may be considered as "the same size as 'b'", "ηρςφχψ" -- as "g", and "βζξ" seem not to have an analogue in English -- that would be 4 settings; capital letters should all have the same size, that of "b" = that of "B", look at: "ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩ").

Such setting will allow to have this user-defined and seems as an ok solution in many cases.

**Comment by kevinbarabash** [comment]
@YakovL for `αβ...` we have the metrics because we support `\alpha\beta...`.  So we should be able to display the greek characters as is in both math mode and text mode.  For Cyrillic though we'd have to do some more work.

**Comment by YakovL** [comment]
So, could you explain, what exactly is needed to display one exact letter (say, 'в') of the same size as 'v'?

**Comment by kevinbarabash** [comment]
We'd need some sort of mapping from Cyrillic characters to latin characters as well as some code that knows to look up the latin character metrics when it encounters Cyrillic characters.

**Comment by episodeyang** [comment]
Hi, I'm having problem with CJK characters. would google font suffice for this purpose? Also the CJK font used for XeLaTeX can be found here:

http://tex.stackexchange.com/questions/53599/standard-cjk-fonts-for-latex

**Comment by episodeyang** [comment]
Hi, I just want to raise your attention on this issue. (honestly I won't be able to contribute to the solution, but as a user I want to add to the signal).  The CJK support is very important for a large range of users. If possible I hope this can be addressed.

**Comment by kevinbarabash** [comment]
@episodeyang with CJK I was worried that we'd need metrics for each glyph, but according to https://en.wikipedia.org/wiki/Monospaced_font it appears that CJK fonts are monospaced which means that we could probably handle them without too much trouble.  We might even be able to do some measurement of glyphs so that we could use whatever font the user is using.  Of course we'd only want to do this measurement once for a couple of glyphs at the start so that we don't affect KaTeX's speed.

**Comment by cben** [comment]
You mean measure at run time?
Currently KaTeX output is a browser-independent pure function, right?
That's a very nice property, for server-side rendering, archival etc.

I suppose in-browser measurement is unavoidably attractive to support whatever fonts users have installed.  Alternatives I can think of:

- Allow any unmeasured characters, taking a guess — but also *enforce* the guessed size (or at least height) via CSS?
- Have some build process that measures given webfonts into metrics KaTeX could use, producing fixed output that expects those webfonts to be served?

**Comment by kevinbarabash** [comment]
@cben good point.  I guess we could use the height of an uppercase latin as our guess for the height of CJK and see how that works.  Adding a measurement step to build process would increase accuracy but it would make the library harder to use.

**Comment by episodeyang** [comment]
that sounds very reasonable to me.

**Comment by episodeyang** [comment]
@kevinbarabash Hi Kevin!
I just want to bump this issue up to your notice. This is so far the biggest hurdle for me to use this in my app. Lack of CJK support would be the biggest reason to go for alternative options.

Thanks for your work, and hopefully these heuristics can be implemented soon.

**Comment by episodeyang** [comment]
> @episodeyang with CJK I was worried that we'd need metrics for each glyph, but according to https://en.wikipedia.org/wiki/Monospaced_font it appears that CJK fonts are monospaced which means that we could probably handle them without too much trouble. We might even be able to do some measurement of glyphs so that we could use whatever font the user is using. Of course we'd only want to do this measurement once for a couple of glyphs at the start so that we don't affect KaTeX's speed.

Indeed! CJK is monospaced ! (in fact doublespaced.) So only a **very** simple measurement is needed.

**Comment by kevinbarabash** [comment]
@episodeyang there is support for CJK now inside `\text` blocks.

**Comment by episodeyang** [comment]
Thanks Kevin! Are there support in mathmode in general? :D

**Comment by kevinbarabash** [comment]
@episodeyang could you provide some example TeX code and expected output?

**Comment by sassanh** [comment]
@kevinbarabash even though:

> Currently KaTeX output is a browser-independent pure function, right?
> That's a very nice property, for server-side rendering, archival etc.

and this is indeed great to have KaTeX a pure function, I guess it'd be useful to support in browser glyph measurement for unknown glyphs/fonts. It can save measurements for each glyph in local storage in user's browser to reduce further render time. This way we can support all fonts and all unicode characters. It can be an optional config of KaTeX.

**Comment by kevinbarabash** [comment]
I think we could measure as a fallback and fail if we're running in node.
